### PR TITLE
Remove hard_limit_for_bookkeeping

### LINKED ISSUE #74401: Fail to get_free_region in thread_final_regions due to low memory for mark array might lead to AV.
While our logic guarantee a free region is available in the free region list before `thread_final_regions`, in case we are running out of memory to commit for the mark array, the call to `init_table_for_region` called in `get_free_region` may fail and therefore return `nullptr`.

To reproduce the issue, we can run `GCPerfSim` with these parameters: 

```
GCPerfSim.dll -tc 6 -tagb 100.0 -tlgb 0.2 -lohar 0 -pohar 0 -sohsi 10 -lohsi 0 -pohsi 0 -sohsr 100-4000 -lohsr 102400-204800 -pohsr 100-4000 -sohpi 10 -lohpi 0 -sohfi 0 -lohfi 0 -pohfi 0 -allocType reference -testKind time
```

and these environment variables:

```
COMPlus_GCHeapCount=6
COMPlus_GCHeapHardLimit=1C200000
COMPlus_GCServer=1
```

On this commit 5f80c24a6600e9d52a84f22d7b6088fe63180cee with a simple change to replace this line in `sufficient_space_regions`

```c++
return check_against_hard_limit (end_space_required);
```
 
to

```c++
return true;
```

To make this issue easier to identify, we can add a check after `get_free_region` in `thread_final_regions` as follow. Change 

```c++
start_region = get_free_region (gen_idx);
```

to

```c++
start_region = get_free_region (gen_idx);
if (start_region == nullptr) { FATAL_GC_ERROR(); }
```

The process will fail at the spot. Without the check, the `nullptr` might be threaded into the generation, and will lead to bad consequences (such as an AV).

### PR DESCRIPTION
**Pull request by PeterSolMS**
As it turns out, it's not so easy to set aside a hard limit for the GC's bookkeeping, because some regions will be partially comitted and so will need more bookkeeping information as a percentage.

So this removes the hard_limit_for_bookkeeping and associated fields, and instead commits memory for the mark array eagerly, even if BGC is not in progress.

### CODE REVIEW

File: src/coreclr/gc/gcpriv.h [hunk 1]
```diff
@@ -4132,10 +4126,6 @@ class gc_heap
     size_t committed_by_oh_per_heap[total_oh_count];
 #endif // _DEBUG && MULTIPLE_HEAPS
 
-    // This is what GC uses for its own bookkeeping.
-    PER_HEAP_ISOLATED
-    size_t current_total_committed_bookkeeping;
```
Comment (cshung) [inline src/coreclr/gc/gcpriv.h:4137]: This field was there before I introduced `heap_hard_limit_for_bookkeeping`.

File: src/coreclr/gc/gc.cpp [hunk 2]
```diff
@@ -20340,20 +20285,17 @@ bool gc_heap::try_get_new_free_region()
 bool gc_heap::init_table_for_region (int gen_number, heap_segment* region)
 {
 #ifdef BACKGROUND_GC
-        if (is_bgc_in_progress())
+        dprintf (GC_TABLE_LOG, ("new seg %Ix, mark_array is %Ix",
+            heap_segment_mem (region), mark_array));
+        if (((region->flags & heap_segment_flags_ma_committed) == 0) &&
+            !commit_mark_array_new_seg (__this, region))
```
Comment (cshung) [inline src/coreclr/gc/gc.cpp:20291]: Shall we guard this under `gc_can_use_concurrent`?

**Review by cshung — COMMENTED** [review]
I liked this idea more than trying to set aside memory for bookkeeping. The one thing I am not sure about is that now we are committing the mark array for every region. If background GC is happening anyway, it might be fine because we will be committing them anyway, but what if we don't?

**Comment by Maoni0** [comment]
> If background GC is happening anyway, it might be fine because we will be committing them anyway, but what if we don't?

that _is_ the drawback with doing this. but if concurrent GC is enabled, it's very likely we'll need to do a BGC at some point so it's not too bad.

File: src/coreclr/gc/gc.cpp (hunk 2 above)
Comment (Maoni0) [reply src/coreclr/gc/gc.cpp:20291]: `commit_mark_array_new_seg` shouldn't do anything if concurrent is disabled; although I'm looking at the code, we are not initializing `background_saved_highest_address` in `init_gc_heap` which we should. @PeterSolMS could you please add this (ie, init `background_saved_highest_address` and `background_saved_lowest_address` to 0 in `init_gc_heap` like we do with WKS GC)? we could still guard this with `gc_can_use_concurrent` just to indicate our intention but it's not essential.

File: src/coreclr/gc/gcpriv.h (hunk 1 above)
Comment (PeterSolMS) [reply src/coreclr/gc/gcpriv.h:4137]: Oops, ok, I'll put it back...

File: src/coreclr/gc/gc.cpp (hunk 2 above)
Comment (PeterSolMS) [reply src/coreclr/gc/gc.cpp]: I just pushed changes with the initialization of `background_saved_lowest_address` and `background_saved_highest_address`, and the restoration of  `current_total_committed_bookkeeping`.

**Comment by PeterSolMS** [comment]
I think I found a hole in our logic - `clear_region_info` (called by `return_free_region`) will *decommit* the mark array in high memory load. When `get_free_region` tries to commit again, we'll fail and get the same AV as before. Perhaps this means we'll have to keep the mark array committed, i.e. disable the logic to decommit the mark array in high memory load?

**Comment by PeterSolMS** [comment]
With this change, we get much closer to actually exhausting the available memory in hard limit cases than before.

In a test case with hard limit set to 0x13c1d000 = 471,859,200, we only get to 331,468,800 committed bytes before running out of memory. The reason is that the memory set aside for bookkeeping is not sufficient, so we run into a situation where the mark array for a region can not be committed.

With this change, we get to 471,842,816 committed bytes before running out. "!dumpheap -stat" reports 462,803,767 bytes in objects, of which 10,684,992 are in free objects, so that we have a total of 452,118,775 bytes in useful objects. 6,078,464 are used for bookkeeping purposes (mainly for the mark array).

**Review by Maoni0 — APPROVED** [review]

**Comment by frankbuckley** [comment]
> With this change, we get much closer to actually exhausting the available memory in hard limit cases than before.
> 
> In a test case with hard limit set to 0x13c1d000 = 471,859,200, we only get to 331,468,800 committed bytes before running out of memory...
> 
> With this change, we get to 471,842,816 committed bytes before running out...

Will this be backported to 7.0?

**Pull request merged**